pkg/scheduler/plugins/tdm/tdm.go

```diff
@@ -0,0 +1,315 @@
+		}()
+	}
+
+	// tdm plugin just handle revocable node
+	predicateFn := func(task *api.TaskInfo, node *api.NodeInfo) error {
+		if node.RevocableZone == "" {
+			return nil
+		}
+
+		if err := bp.availableRevocableZone(node.RevocableZone); err != nil {
+			return fmt.Errorf("plugin %s predicates %w", bp.Name(), err)
+		}
+
+		if !task.Preemptable {
+			msg := fmt.Sprintf("task %s/%s is not preemptable task, skip to schedule to node %s", task.Namespace, task.Name, node.Name)
+			return fmt.Errorf("plugin %s predicates %s", bp.Name(), msg)
+		}
+
+		klog.V(4).Infof("TDM filter for Task %s/%s on node %s pass.", task.Namespace, task.Name, node.Name)
+		return nil
+	}
+
+	// tdm plugin just handle revocable node
+	nodeOrderFn := func(task *api.TaskInfo, node *api.NodeInfo) (float64, error) {
+		score := 0.0
+
+		if node.RevocableZone == "" {
+			return score, nil
+		}
+
+		if err := bp.availableRevocableZone(node.RevocableZone); err != nil {
+			klog.V(4).Infof("TDM not available %s", err)
+			return score, err
+		}
+
+		if !task.Preemptable {
+			klog.V(4).Infof("TDM task %s/%s is not preemptable task, skip to schedule to node %s", task.Namespace, task.Name, node.Name)
+			return score, nil
+		}
+
+		score = float64(v1alpha1.MaxNodeScore)
+
+		klog.V(4).Infof("TDM score for Task %s/%s on node %s is: %v", task.Namespace, task.Name, node.Name, score)
+		return score, nil
+	}
+
+	preemptableFn := func(preemptor *api.TaskInfo, preemptees []*api.TaskInfo) []*api.TaskInfo {
+		if preemptor.Preemptable {
+			klog.V(4).Infof("task %s/%s is preemptable, do nothing skip", preemptor.Namespace, preemptor.Name)
+			return preemptees
+		}
+
+		var victims []*api.TaskInfo
+
+		// find preemptable tasks which appear on none revocable node
+		for _, task := range preemptees {
+			if !task.Preemptable {
+				continue
+			}
+
+			node, ok := ssn.Nodes[task.NodeName]
+			if !ok {
+				continue
+			}
+
+			if node.RevocableZone != "" {
+				continue
+			}
+
+			victims = append(victims, task)
+		}
+
+		klog.V(4).Infof("TDM victims are %+v", victims)
+
+		return victims
+	}
+
+	victimsFn := func() []*api.TaskInfo {
+		if lastEvictAt.Add(bp.evictPeriod).After(time.Now()) {
+			klog.V(4).Infof("next evict time at %v", lastEvictAt)
+			return nil
+		}
+
+		// 1. find preemptable task on timeout revocable zone node
+		// 2. if there is active revocable zone find preemptable task on none revocable node
+
+		active := make(map[string]struct{})
+		victims := make([]*api.TaskInfo, 0)
+		for rz := range bp.revocableZone {
+			if err := bp.availableRevocableZone(rz); err != nil {
+				// rz disactive, then evict preemptable tasks from the revocable node
+				victims = append(victims, bp.revocableNodePreemptableTask(rz, ssn)...)
+			} else {
+				active[rz] = struct{}{}
+			}
+		}
+
+		// if there are some active rz, then evict preemptable tasks from all the none revocable node
+		victims = append(victims, bp.noneRevocableNodePreemptableTask(ssn)...)
```

Comment:
It seems there is no check if has active rz. If there is no active rz, the preemtable pod on non revocable zone will be evicted again and again.

Reply:
Yes, fixed.